packages/core/src/synthesis/Synthesizer.ts

```diff
@@ -624,143 +671,243 @@ export class Synthesizer {
       log.debug(
         `Warning: cannot find a ${stmtType} statement with name ${name}.`
       );
+      log.debug(`Available statements:\n${prettySubstance(this.currentProg)}`);
+      return undefined;
     }
   };
 
-  generateType = (typeName?: Identifier): Decl => {
-    // pick a type
-    let typeCons: TypeConsApp;
-    if (typeName) {
-      typeCons = nullaryTypeCons(typeName);
-    } else {
-      const type: TypeDecl = this.choice(
-        this.cxt.candidates.types.toArray().map(([, b]) => b)
-      );
-
-      typeCons = applyTypeDecl(type);
-    }
+  generateDecl = (type: TypeDecl, ctx: SynthesisContext): Decl => {
+    const typeCons = applyTypeDecl(type);
+    const { res: name } = this.generateID(ctx, typeCons.name);
     const stmt: Decl = {
       tag: "Decl",
       nodeType: "SyntheticSubstance",
       children: [],
       type: typeCons,
-      name: this.cxt.generateID(typeCons.name),
+      name,
     };
-    this.cxt.appendStmt(stmt);
     return stmt;
   };
 
-  generatePredicate = (pred?: PredicateDecl): ApplyPredicate => {
-    if (!pred) {
-      pred = this.choice(
-        this.cxt.candidates.predicates.toArray().map(([, b]) => b)
-      );
-    }
-    const args: SubPredArg[] = this.generatePredArgs(pred.args);
-    const stmt: ApplyPredicate = applyPredicate(pred, args);
-    this.cxt.appendStmt(stmt);
+  generateDeclFromType = (
+    typeCons: TypeConsApp,
+    ctx: SynthesisContext
+  ): Decl => {
+    const { res: name } = this.generateID(ctx, typeCons.name);
+    const stmt: Decl = {
+      tag: "Decl",
+      nodeType: "SyntheticSubstance",
+      children: [],
+      type: typeCons,
+      name,
+    };
     return stmt;
   };
 
-  generateFunction = (func?: FunctionDecl): Bind => {
-    if (!func) {
-      func = this.choice(
-        this.cxt.candidates.functions.toArray().map(([, b]) => b)
-      );
-    }
-    const args: SubExpr[] = this.generateArgs(func.args);
+  generatePredicate = (
+    pred: PredicateDecl,
+    ctx: SynthesisContext
+  ): WithStmts<ApplyPredicate> => {
+    const { res, stmts }: WithStmts<SubPredArg[]> = this.generatePredArgs(
+      pred.args,
+      ctx
+    );
+    const p: ApplyPredicate = applyPredicate(pred, res);
+    return { res: p, stmts };
+  };
+
+  generateFunction = (
+    func: FunctionDecl,
+    ctx: SynthesisContext
+  ): WithStmts<Bind> => {
+    const { res: args, stmts: decls }: WithStmts<SubExpr[]> = this.generateArgs(
+      func.args,
+      ctx
+    );
     const rhs: ApplyFunction = applyFunction(func, args);
+    // find the `TypeDecl` for the output type
     const outputType = func.output.type as TypeConstructor;
+    // NOTE: the below will bypass the candidate list and generate a new decl, to follow the config more strictly:
+    // const outputTypeDecl: TypeDecl | undefined = ctx.candidates.types.find(
+    //   (decl, typeName) => typeName === outputType.name.value
+    // );
```

Comment:
nit: in practice, is there a reason to keep this?

Reply:
Deleting the code as it's outdated anyway. Still keeping the `NOTE:` so we are aware of this and may fix this later.